File: app/web.py
"""Infraestrutura de apresentacao: Jinja2, filtros e helpers de template."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote, urlencode

from fastapi import Request
from fastapi.responses import RedirectResponse
from fastapi.templating import Jinja2Templates
from jinja2 import pass_context
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app import modulos
from app.config import RAIZ, RODAPE_INSTITUCIONAL, VERSAO, obter_config
from app.modelos.estados import COLUNAS_KANBAN, ROTULO_ESTADO_TELA
from app.seguranca import cookie_seguro
from app.servicos import (
    auditoria,
    autenticacao,
    datas_br,
    epi_estoque,
    identificacao,
)
from app.servicos.rbac import UsuarioAtual

templates = Jinja2Templates(directory=str(RAIZ / "app" / "templates"))

templates.env.filters["data"] = lambda d: datas_br.numerica(d) if d else ""
templates.env.filters["data_extenso"] = lambda d: datas_br.por_extenso(d) if d else ""
templates.env.filters["momento"] = lambda m: datas_br.local_formatado(m)
# O CNPJ e guardado so em digito (a coluna e String(14) e o servico normaliza),
# porque pontuacao nao e dado; ela e leitura. Quatorze digitos corridos numa
# celula sao exatamente o que ninguem confere contra a nota fiscal, entao a
# pontuacao volta AQUI, na saida — filtro e nao coluna, pelo mesmo motivo de
# `estado`: um caminho unico do dado ate a tela.
templates.env.filters["cnpj"] = epi_estoque.formatar_cnpj
# `ROTULO_ESTADO_TELA` e nao `ROTULO_ESTADO`: o mapa ASCII continua sendo o que
# a mensagem de servico e a descricao da trilha escrevem (o texto delas entra no
# digest da auditoria), e este filtro e o caminho unico do estado ate a tela —
# pilula, cartao do kanban, seletor de destino e filtro de lista.
templates.env.filters["estado"] = lambda e: ROTULO_ESTADO_TELA.get(e, e or "")
# O tipo de evento tambem e chave de banco, e pela mesma razao ganha filtro em
# vez de mapa por template: a trilha aparece em quatro telas de tres modulos.
templates.env.filters["evento"] = auditoria.rotulo_evento
templates.env.globals["VERSAO"] = VERSAO
templates.env.globals["RODAPE"] = RODAPE_INSTITUCIONAL
templates.env.globals["COLUNAS_KANBAN"] = COLUNAS_KANBAN
# O seletor da lateral precisa saber por onde se entra em cada módulo, e isso
# depende do usuário. Vai como global pelo mesmo motivo de COLUNAS_KANBAN: e
# constante de apresentacao, nao contexto de tela, e a decisao de qual e a
# primeira tela visivel tem de ser a mesma que `/inicio` toma — duas respostas
# para "onde comeca o modulo de EPI" divergiriam no primeiro item novo.
templates.env.globals["porta_de_entrada"] = modulos.porta_de_entrada

# A barra de forca da senha (primeiro acesso e troca) mede a politica de
# verdade, e nao uma nocao propria de "senha forte". Para isso ela precisa dos
# mesmos numeros e da mesma lista que `autenticacao.politica_de_senha` usa —
# reescrever qualquer um dos dois no JavaScript criaria uma segunda regra, que
# divergiria da primeira no dia em que alguem mexesse so num lado. Vai como
# global do Jinja pelo mesmo motivo de COLUNAS_KANBAN: e constante de
# apresentacao, nao contexto de tela, e nenhuma rota precisa lembrar de passar.
# A validacao que vale continua sendo a do servidor: aqui e so o aviso previo.
templates.env.globals["POLITICA_SENHA"] = {
    "minimo": autenticacao.TAMANHO_MINIMO_SENHA,
    "triviais": sorted(autenticacao.SENHAS_TRIVIAIS),
}


@pass_context
def _identificar(contexto, alvo, vazio: str = identificacao.SEM_SERVIDOR):
    """RN-19 no template sem a tela precisar carregar nada.

    E global de contexto, e nao parametro de `pagina()`, exatamente para a tela
    nova nao poder esquecer: `usuario` e `request` ja estao no contexto de toda
    pagina e de todo fragmento, entao o template escreve `identificar(x)` e
    pronto. Nao havendo `usuario` no contexto (tela de erro antes do login), a
    decisao cai no lado seguro — identificador opaco.
    """
    return identificacao.identificar(
        alvo,
        contexto.get("usuario"),
        identificacao.semente_de(contexto.get("request")),
        vazio=vazio,
    )


templates.env.globals["identificar"] = _identificar


@pass_context
def _texto_livre(contexto, texto, sobre=None, vazio: str = identificacao.SEM_SERVIDOR):
    """A irma de `identificar` para a frase que o sistema gravou.

    Global de contexto pelo mesmo motivo, e com mais razao: `descricao` de
    trilha e de pendencia e renderizada em sete telas de quatro modulos, e a que
    esquecesse de chamar seria justamente o vazamento — foi assim que o nome com
    SIAPE continuou saindo na ficha da requisicao depois de a RN-19 estar de pe
    em toda coluna de nome do sistema.
    """
    return identificacao.texto_livre(
        texto,
        contexto.get("usuario"),
        sobre=sobre,
        vazio=vazio,
    )


templates.env.globals["texto_livre"] = _texto_livre


@pass_context
def _nonce(contexto) -> str:
    """A marca do `<script>` embutido desta resposta.

    E global de contexto pelo mesmo motivo de `identificar`: `request` ja esta no
    contexto de toda pagina, entao o template escreve `nonce="{{ nonce() }}"` e
    pronto — nenhuma rota precisa lembrar de passar nada. Quem escreve o valor e
    o middleware `CabecalhosDeSeguranca`, uma vez por resposta, e o mesmo valor
    vai para o cabecalho `Content-Security-Policy`.

    Fora de requisicao com middleware (um `TemplateResponse` montado a mao num
    teste) devolve vazio: sem cabecalho de CSP nao ha o que casar, e uma marca
    inventada aqui daria a impressao de que o script esta autorizado quando o
    que o autoriza e o cabecalho.
    """
    estado = getattr(contexto.get("request"), "state", None)
    return getattr(estado, "nonce_csp", "") if estado is not None else ""


templates.env.globals["nonce"] = _nonce


def _sino(
    usuario: UsuarioAtual | None, s_requisicao: Session | None = None
) -> tuple[int, int]:
    """Contagem do sino: (abertas, atrasadas). Falha em silencio - o sino nunca
    pode derrubar a tela que ele decora.

    **Conta pela sessao da requisicao quando ela existe.** Antes o sino abria
    sempre uma sessao propria, e isso era o deadlock consigo mesmo que o
    `banco.py` avisa em `_transacao_imediata`: a sessao da requisicao ja tomou o
    lock de escrita com `BEGIN IMMEDIATE`, entao a segunda conexao esperava o
    `busy_timeout` inteiro e terminava em "database is locked" — capturado aqui
    pelo `except`. Resultado: **toda tela HTML demorava ~5,6 s a mais e o sino
    marcava zero**, sempre, para todo mundo. O silencio do `except` e que fez
    isso durar: o sino nao pode derrubar a tela, mas tambem nao pode mentir em
    silencio, e ele mentia.

    O `no_autoflush` e deliberado: contar pendencia e leitura de decoracao de
    cabecalho e nao pode arrastar para o banco um flush do que a rota ainda
    estava montando.
    """
    if usuario is None:
        return 0, 0
    try:
        from app.servicos.pendencias import contar_abertas

        if s_requisicao is not None:
            with s_requisicao.no_autoflush:
                return contar_abertas(s_requisicao, usuario)

        # fora de requisicao (tela de erro, com a sessao ja encerrada) nao ha
        # lock para disputar: abrir uma curta e o certo.
        from app.banco import sessao

        with sessao() as s:
            return contar_abertas(s, usuario)
    except Exception:  # pragma: no cover - defensivo
        return 0, 0


def pagina(
    request: Request,
    template: str,
    usuario: UsuarioAtual | None = None,
    **contexto: Any,
):
    from app.dependencias import sessao_da_requisicao
    from app.servicos.busca import pode_buscar
    from app.servicos.pendencias import pode_ver as pode_ver_pendencias

    cfg = obter_config()
    # o sino e um link para /pendencias: quem nao pode abrir a tela nao ganha o
    # link. Menu que oferece porta trancada e a mesma mentira do item de modulo.
    sino_visivel = pode_ver_pendencias(usuario)
    abertas, atrasadas = (
        _sino(usuario, sessao_da_requisicao(request)) if sino_visivel else (0, 0)
    )
    origem = request.cookies.get(modulos.COOKIE_MODULO)
    navegacao = modulos.navegacao(request.url.path, usuario, origem=origem)
    dados = {
        "request": request,
        "usuario": usuario,
        "avisos_config": cfg.inseguro,
        "ambiente": cfg.ambiente,
        "sino_visivel": sino_visivel,
        "sino_abertas": abertas,
        "sino_atrasadas": atrasadas,
        # a caixa do cabeçalho, pela mesma regra do sino: quem não abre nenhum
        # bloco de `/buscar` não ganha o campo. Vai daqui, e não de um
        # `usuario.pode(...)` no template, porque a lista de blocos cresce a cada
        # módulo — a caixa tem de acompanhar sozinha, sem ninguém lembrar.
        "busca_visivel": pode_buscar(usuario),
        **navegacao,
        **contexto,
    }
    resposta = templates.TemplateResponse(request, template, dados)
    # Toda tela de módulo anota de onde a pessoa está vindo, para que a lateral
    # não a expulse do módulo quando ela abre um cadastro da base. Só grava
    # quando muda: um `Set-Cookie` em toda resposta HTML atrapalharia cache
    # intermediário sem trocar informação nenhuma.
    lembrar = navegacao["modulo_a_lembrar"]
    if lembrar and lembrar != origem:
        resposta.set_cookie(
            modulos.COOKIE_MODULO,
            lembrar,
            httponly=True,
            samesite="lax",
            path="/",
            max_age=cfg.sessao_horas * 3600,
            # mesma decisao, mesma funcao: era aqui que o calculo do `Secure`
            # estava DUPLICADO, e duplicata e como o defeito de `/login` chegou
            # tambem ao cookie de modulo.
            secure=cookie_seguro(request),
        )
    return resposta


def marcar_download(resposta, request: Request):
    """Devolve ao navegador a marca que a página mandou junto com o pedido.

    Um download não dispara evento nenhum na página que o pediu: o "Gerar
    PDF" chamava o LibreOffice por até 180 s e o botão continuava clicável,
    sem sinal de que algo estava acontecendo. A página manda `?baixar=<token>`
    no clique; esta função grava o mesmo token num cookie que a página observa
    (`csso.js`, bloco 4) e, ao vê-lo, religa o botão. Sem `baixar` na query —
    link colado, teste, quem está sem JavaScript — não grava nada.

    `httponly=False` de propósito: o cookie existe para ser LIDO pela página.
    Não carrega informação nenhuma além do token que a própria página gerou,
    e morre em 60 s.
    """
    # a rota chamada direto (teste, ferramenta) nao tem requisicao: nada a marcar
    token = request.query_params.get("baixar", "") if request is not None else ""
    if token and re.fullmatch(r"[A-Za-z0-9]{1,64}", token):
        resposta.set_cookie(
            "csso_baixou", token, max_age=60, path="/", httponly=False, samesite="lax",
            secure=cookie_seguro(request),
        )
    return resposta


def fragmento(request: Request, template: str, **contexto: Any):
    return templates.TemplateResponse(request, template, {"request": request, **contexto})


# =====================================================================
# Paginacao de lista ja materializada
# =====================================================================
POR_PAGINA = 50


@dataclass(frozen=True)
class Recorte:
    """Uma pagina de uma lista, com a escala que a tela precisa dizer.

    O custo que isto resolve e o que esta medido no achado que originou a
    paginacao: `/servidores` com 975 linhas de seis colunas e um `title` por
    celula devolve ~1MB de HTML. E tamanho de PAGINA, nao de consulta — o sistema
    declara `lazy="selectin"` em 86 relacionamentos e as listas nao tem N+1.
    Cinquenta `<tr>` no lugar de 975 e a correcao inteira.

    **`pagina` E APARADA**, e essa e a diferenca para a paginacao cega de
    `/auditoria`. Aqui o total e conhecido, entao `?pagina=999` numa lista de
    tres paginas nao pode devolver uma tela vazia sem explicacao: ela volta para
    a pagina 3. Numero zero ou negativo volta para a 1 — sem isto o `itens[-100:]`
    do Python devolveria silenciosamente o FIM da lista, que e o pior resultado
    possivel: uma pagina plausivel e errada. Em `/auditoria` a contagem nao existe
    de proposito (contar doze mil eventos a cada abertura da trilha sairia caro),
    e por isso la a passagem do fim continua sendo uma pagina vazia com o aviso
    escrito na tela.
    """

    itens: list
    pagina: int
    por_pagina: int
    total: int


def _aparar(pagina, total: int, por_pagina: int) -> int:
    ultima = max(1, (total + por_pagina - 1) // por_pagina)
    return min(max(int(pagina or 1), 1), ultima)


def recortar(itens: list, pagina: int = 1, por_pagina: int = POR_PAGINA) -> Recorte:
    """Recorta uma lista JA MATERIALIZADA.

    **Existe porque nem todo filtro cabe na consulta.** `/servidores` casa a
    busca por `casa_a_busca`, que aplica a RN-19 linha a linha (o SIAPE acha para
    todo mundo, o nome so para quem ja podia le-lo); `/epis/fichas` filtra sobre
    uma linha derivada de tres consultas; `/laudos` casa por unidade e por
    subscritor, que sao relacionamentos. Empurrar o LIMIT para o banco antes
    desses filtros pagina o conjunto ERRADO — a pagina 1 sairia com quatro linhas
    porque as outras 46 foram recortadas depois.

    Onde o filtro CABE na consulta, o certo e `recortar_consulta`, abaixo: ali o
    banco devolve so as 50 linhas, e nao as 975.
    """
    total = len(itens)
    pagina = _aparar(pagina, total, por_pagina)
    inicio = (pagina - 1) * por_pagina
    return Recorte(itens[inicio : inicio + por_pagina], pagina, por_pagina, total)


def recortar_consulta(
    s: Session, consulta, pagina: int = 1, por_pagina: int = POR_PAGINA
) -> Recorte:
    """Recorta no BANCO: uma contagem e um LIMIT/OFFSET.

    Serve a lista cujo filtro inteiro ja esta na consulta — `/certificados` e o
    caso: situacao, treinamento e chave de validacao sao todos `where`. Ali
    materializar 975 registros para jogar fora 925 seria trabalho que o banco faz
    melhor.

    O `order_by(None)` na contagem nao e zelo: ordenar uma subconsulta de COUNT e
    trabalho puro (o SQLite chega a montar um indice temporario para isso) e nao
    muda o numero.
    """
    total = s.execute(
        select(func.count()).select_from(consulta.order_by(None).subquery())
    ).scalar_one()
    pagina = _aparar(pagina, total, por_pagina)
    itens = list(
        s.execute(consulta.limit(por_pagina).offset((pagina - 1) * por_pagina)).scalars()
    )
    return Recorte(itens, pagina, por_pagina, total)


def numero_da_pagina(request: Request) -> int:
    """O `?pagina=` da URL, ou 1.

    Sai de `query_params` e nao da assinatura da rota — em toda rota deste
    sistema `pagina` e o nome do ajudante que RENDERIZA a tela, e um parametro
    com esse nome o sombrearia dentro da funcao. E o mesmo motivo pelo qual
    `processos._filtro` ja lia a pagina dali.
    """
    bruto = request.query_params.get("pagina", "")
    return int(bruto) if bruto.isdigit() else 1


def salvar_com_diff(
    s,
    usuario: UsuarioAtual,
    modelo,
    registro_id: int,
    campos: dict,
    *,
    permissao: str,
    rotulo: str,
    volta: str,
):
    """Edição de catálogo: exige a permissão, grava e audita campo a campo.

    Vive aqui, e não em cada arquivo de rota, porque a regra é a mesma em todo
    catálogo do sistema — o que muda de um para outro é só a permissão exigida.
    Duplicar isto foi como o menu passou a mentir na 1.6.0.
    """
    usuario.exigir(permissao)
    registro = s.get(modelo, registro_id)
    if registro is None:
        return RedirectResponse(volta, status_code=303)
    antes = {campo: getattr(registro, campo) for campo in campos}
    for campo, valor in campos.items():
        setattr(registro, campo, valor)
    auditoria.registrar_diferencas(
        s,
        entidade=modelo.__tablename__,
        entidade_id=registro.id,
        antes=antes,
        depois={campo: getattr(registro, campo) for campo in campos},
        usuario=usuario,
    )
    s.commit()
    # codificado: `rotulo` é o nome do catálogo e chega ao `Location` dentro da
    # mensagem. Sem isso, um `&` no rótulo encerra a query string ali e o recado
    # some — a mesma correção que os helpers `_aviso` das rotas receberam.
    parametros = urlencode({"mensagem": f"{rotulo} atualizado."}, quote_via=quote)
    separador = "&" if "?" in volta else "?"
    return RedirectResponse(f"{volta}{separador}{parametros}", status_code=303)

File: nuvem/src/web.ts
/**
 * Infraestrutura de apresentação: Nunjucks, filtros e helpers de template.
 * Porte de `app/web.py`.
 *
 * **Por que Nunjucks.** É o Jinja2 do JavaScript: a mesma sintaxe de bloco,
 * herança, macro, `include` e filtro. Os 64 templates do sistema Python vêm
 * para `templates/` quase como estavam; o que muda está listado em
 * `docs/TEMPLATES.md` (o `.items()` de dicionário, `namespace()`, `|tojson`...).
 *
 * Nunjucks renderiza de forma síncrona: tudo que a tela mostra tem de estar
 * carregado ANTES de `pagina()` — não há consulta preguiçosa dentro do
 * template, como havia com o `lazy="selectin"` do SQLAlchemy. Quem porta uma
 * rota carrega as relações que o template lê (`with:` do Drizzle).
 */
import nunjucks from "nunjucks";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { existsSync } from "node:fs";
import type { Ctx } from "./nucleo/contexto.js";
import { obterConfig, RODAPE_INSTITUCIONAL, VERSAO } from "./config.js";
import { cookieSeguro } from "./nucleo/seguranca.js";
import { getCookie, setCookie } from "hono/cookie";
import type { UsuarioAtual } from "./servicos/rbac.js";
import { COLUNAS_KANBAN } from "./dominio/estados.js";
import * as datas_br from "./servicos/datas_br.js";
import { rotulo_evento } from "./servicos/auditoria.js";
import * as identificacao from "./servicos/identificacao.js";

/**
 * Onde estão os templates. Na função do Netlify o código vira um bundle em
 * outro diretório e os templates vão junto por `included_files`; localmente
 * ficam em `nuvem/templates`. Procura-se nos dois.
 */
function pastaDeTemplates(): string {
  const candidatos = [
    process.env.CSSO_TEMPLATES,
    path.resolve(process.cwd(), "templates"),
    path.resolve(process.cwd(), "nuvem/templates"),
    path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../templates"),
  ].filter(Boolean) as string[];
  for (const c of candidatos) if (existsSync(path.join(c, "base.html"))) return c;
  return candidatos[1]!;
}

export const ambiente = new nunjucks.Environment(
  new nunjucks.FileSystemLoader(pastaDeTemplates(), { noCache: process.env.NODE_ENV === "development" }),
  { autoescape: true, throwOnUndefined: false, trimBlocks: false, lstripBlocks: false },
);

// ---------------------------------------------------------------------
// Compatibilidade Jinja -> Nunjucks: filtros que o Jinja tem e o Nunjucks não
// ---------------------------------------------------------------------
ambiente.addFilter("tojson", (v: unknown) =>
  new nunjucks.runtime.SafeString(
    JSON.stringify(v ?? null)
      .replace(/</g, "\\u003c")
      .replace(/>/g, "\\u003e")
      .replace(/&/g, "\\u0026")
      .replace(/'/g, "\\u0027"),
  ),
);
ambiente.addFilter("items", (obj: Record<string, unknown> | Map<unknown, unknown> | null) => {
  if (!obj) return [];
  if (obj instanceof Map) return [...obj.entries()];
  return Object.entries(obj);
});
ambiente.addFilter("selectattr", (lista: any[], attr: string, teste?: string, valor?: unknown) =>
  (lista ?? []).filter((x) => {
    const v = x?.[attr];
    if (teste === undefined) return Boolean(v);
    if (teste === "equalto" || teste === "eq" || teste === "==") return v === valor;
    if (teste === "ne" || teste === "!=") return v !== valor;
    if (teste === "none") return v === null || v === undefined;
    return Boolean(v);
  }),
);
ambiente.addFilter("rejectattr", (lista: any[], attr: string, teste?: string, valor?: unknown) =>
  (lista ?? []).filter((x) => {
    const v = x?.[attr];
    if (teste === undefined) return !v;
    if (teste === "equalto" || teste === "eq" || teste === "==") return v !== valor;
    if (teste === "none") return !(v === null || v === undefined);
    return !v;
  }),
);
ambiente.addFilter("map", (lista: any[], ...args: any[]) => {
  // Jinja: map(attribute='x') -> o Nunjucks passa kwargs como último objeto
  const ultimo = args[args.length - 1];
  const attr = typeof ultimo === "object" && ultimo?.attribute ? ultimo.attribute : args[0];
  return (lista ?? []).map((x) => x?.[attr]);
});
ambiente.addFilter("max", (lista: number[]) => (lista?.length ? Math.max(...lista) : undefined));
ambiente.addFilter("min", (lista: number[]) => (lista?.length ? Math.min(...lista) : undefined));
ambiente.addFilter("unique", (lista: unknown[]) => [...new Set(lista ?? [])]);
ambiente.addFilter("format", (fmt: string, ...args: unknown[]) => {
  let i = 0;
  return String(fmt).replace(/%(\.(\d+))?([sdif%])/g, (_m, _p, prec, tipo) => {
    if (tipo === "%") return "%";
    const v = args[i++];
    if (tipo === "d" || tipo === "i") return String(Math.trunc(Number(v)));
    if (tipo === "f") return Number(v).toFixed(prec ? Number(prec) : 6);
    return String(v);
  });
});

// testes do Jinja que o Nunjucks não traz
(ambiente as any).addTest("none", (v: unknown) => v === null || v === undefined);
(ambiente as any).addTest("sameas", (v: unknown, outro: unknown) => v === outro);
(ambiente as any).addTest("in", (v: unknown, col: unknown) =>
  Array.isArray(col) ? col.includes(v) : typeof col === "string" ? col.includes(String(v)) : col instanceof Set ? col.has(v) : false,
);
(ambiente as any).addTest("boolean", (v: unknown) => typeof v === "boolean");
(ambiente as any).addTest("sequence", (v: unknown) => Array.isArray(v) || typeof v === "string");

/** Registra um filtro. Os serviços que fornecem filtros chamam isto no carregamento. */
export function filtro(nome: string, f: (...args: any[]) => unknown): void {
  ambiente.addFilter(nome, f);
}
/** Registra um global de template. */
export function global(nome: string, v: unknown): void {
  ambiente.addGlobal(nome, v);
}

global("VERSAO", VERSAO);
global("RODAPE", RODAPE_INSTITUCIONAL);
global("range", (a: number, b?: number, passo = 1) => {
  const [ini, fim] = b === undefined ? [0, a] : [a, b];
  const r: number[] = [];
  for (let i = ini; passo > 0 ? i < fim : i > fim; i += passo) r.push(i);
  return r;
});

// O quadro do kanban é constante de apresentação, não contexto de tela.
global("COLUNAS_KANBAN", COLUNAS_KANBAN);
// `(fim - inicio).days` do Jinja sobre duas `date`: aqui as datas são texto
// 'AAAA-MM-DD' (RN-18), e a subtração vira esta função (`partes/macros.html`).
global("dias_entre", (inicio: string | null, fim: string | null): number | null => {
  if (!inicio || !fim) return null;
  const dia = (s: string) => Date.UTC(Number(s.slice(0, 4)), Number(s.slice(5, 7)) - 1, Number(s.slice(8, 10)));
  return Math.round((dia(String(fim)) - dia(String(inicio))) / 86_400_000);
});
// ---------------------------------------------------------------------
// Datas, evento e RN-19 (porte do núcleo B)
// ---------------------------------------------------------------------
// `date` chega como 'AAAA-MM-DD' e é formatada sem passar por fuso (RN-18);
// `timestamptz` chega como `Date` e é exibido em America/Sao_Paulo.
filtro("data", (d: unknown) => {
  const dia = datas_br.comoData(d);
  return dia ? datas_br.numerica(dia) : "";
});
filtro("data_extenso", (d: unknown) => {
  const dia = datas_br.comoData(d);
  return dia ? datas_br.por_extenso(dia) : "";
});
filtro("momento", (m: Date | string | null | undefined) => datas_br.local_formatado(m ?? null));
// O tipo de evento é chave de banco; o rótulo mora em `auditoria.ROTULO_EVENTO`.
filtro("evento", (codigo: string | null | undefined) => rotulo_evento(codigo));

/** Os kwargs do Nunjucks chegam como último argumento, marcado `__keywords`. */
function separarKwargs(args: unknown[]): [unknown[], Record<string, unknown>] {
  const ultimo = args[args.length - 1] as Record<string, unknown> | undefined;
  if (ultimo && typeof ultimo === "object" && (ultimo as { __keywords?: boolean }).__keywords) {
    const { __keywords: _k, ...kw } = ultimo;
    return [args.slice(0, -1), kw];
  }
  return [args, {}];
}

/**
 * RN-19 no template sem a tela carregar nada: `usuario` e `request` já estão
 * no contexto de toda página e fragmento (Nunjucks chama global com
 * `this.ctx` = contexto). Sem `usuario`, o lado seguro — código opaco.
 */
global("identificar", function (this: { ctx: Record<string, any> }, ...args: unknown[]) {
  const [pos, kw] = separarKwargs(args);
  const vazio = (kw.vazio ?? pos[1] ?? identificacao.SEM_SERVIDOR) as string;
  return identificacao.identificar(
    pos[0] as identificacao.AlvoIdentificavel,
    this.ctx?.usuario ?? null,
    identificacao.semente_de(this.ctx?.request),
    { vazio },
  );
});

/** A irmã de `identificar` para a frase que o sistema gravou (trilha, pendência). */
global("texto_livre", function (this: { ctx: Record<string, any> }, ...args: unknown[]) {
  const [pos, kw] = separarKwargs(args);
  return identificacao.texto_livre(pos[0] as string | null, this.ctx?.usuario ?? null, {
    sobre: (kw.sobre ?? pos[1]) as identificacao.AlvoIdentificavel,
    vazio: (kw.vazio ?? pos[2] ?? identificacao.SEM_SERVIDOR) as string,
  });
});

// Os outros globais que o `web.py` registrava moram no módulo dono deles, que os
// registra ao ser carregado — `web.ts` não os importa para não fechar ciclo de
// importação (esses módulos dependem de `web.ts`, ou dependem de quem depende):
//   - `porta_de_entrada` ............ `src/modulos.ts`
//   - `POLITICA_SENHA` .............. `src/servicos/autenticacao.ts`

/**
 * O `request` que os templates enxergam. Só o que eles de fato leem:
 * caminho, query e cookies — com a mesma forma do Starlette
 * (`request.url.path`, `request.query_params.get('x')`).
 */
export function requestParaTemplate(c: Ctx) {
  const url = new URL(c.req.url);
  return {
    url: { path: url.pathname, query: url.search.replace(/^\?/, ""), scheme: url.protocol.replace(":", "") },
    method: c.req.method,
    path: url.pathname,
    query_params: {
      get: (k: string, padrao: string | null = null) => url.searchParams.get(k) ?? padrao,
      getlist: (k: string) => url.searchParams.getAll(k),
    },
    cookies: { get: (k: string) => getCookie(c, k) ?? null },
    headers: { get: (k: string) => c.req.header(k) ?? null },
  };
}

// ---------------------------------------------------------------------
// Ganchos preenchidos pelos serviços do núcleo quando forem carregados.
// Ficam aqui como pontos de extensão para que `web.ts` não dependa do módulo
// de pendências, busca e navegação (que dependem de `web.ts` de volta).
// ---------------------------------------------------------------------
export interface GanchosDaCasca {
  sino?: (c: Ctx, usuario: UsuarioAtual) => Promise<[number, number]>;
  podeVerPendencias?: (usuario: UsuarioAtual | null) => boolean;
  podeBuscar?: (usuario: UsuarioAtual | null) => boolean;
  navegacao?: (
    caminho: string,
    usuario: UsuarioAtual | null,
    origem: string | null,
  ) => Record<string, unknown> & { modulo_a_lembrar?: string | null };
  cookieModulo?: string;
}
export const ganchos: GanchosDaCasca = {};

export function renderizar(template: string, dados: Record<string, unknown>): string {
  return ambiente.render(template, dados);
}

/**
 * A tela inteira, com a casca (lateral, sino, busca). O `web.pagina` do Python.
 */
export async function pagina(
  c: Ctx,
  template: string,
  usuario: UsuarioAtual | null = null,
  contexto: Record<string, unknown> = {},
  status = 200,
): Promise<Response> {
  const cfg = obterConfig();
  const sinoVisivel = ganchos.podeVerPendencias?.(usuario) ?? false;
  let abertas = 0;
  let atrasadas = 0;
  if (sinoVisivel && usuario && ganchos.sino) {
    try {
      [abertas, atrasadas] = await ganchos.sino(c, usuario);
    } catch {
      // o sino nunca derruba a tela que ele decora
    }
  }
  const cookieModulo = ganchos.cookieModulo ?? "csso_modulo";
  const origem = getCookie(c, cookieModulo) ?? null;
  const url = new URL(c.req.url);
  const navegacao = ganchos.navegacao?.(url.pathname, usuario, origem) ?? {};
  const dados: Record<string, unknown> = {
    request: requestParaTemplate(c),
    usuario,
    avisos_config: cfg.inseguro,
    ambiente: cfg.ambiente,
    sino_visivel: sinoVisivel,
    sino_abertas: abertas,
    sino_atrasadas: atrasadas,
    busca_visivel: ganchos.podeBuscar?.(usuario) ?? false,
    nonce: () => c.get("nonce") ?? "",
    ...navegacao,
    ...contexto,
  };
  const lembrar = navegacao.modulo_a_lembrar as string | null | undefined;
  if (lembrar && lembrar !== origem) {
    setCookie(c, cookieModulo, lembrar, {
      httpOnly: true,
      sameSite: "Lax",
      path: "/",
      maxAge: cfg.sessaoHoras * 3600,
      secure: cookieSeguro(c.req.url),
    });
  }
  return c.html(renderizar(template, dados), status as any);
}

/** Um pedaço de tela (resposta de HTMX), sem casca. */
export function fragmento(c: Ctx, template: string, contexto: Record<string, unknown> = {}, status = 200): Response {
  return c.html(
    renderizar(template, { request: requestParaTemplate(c), nonce: () => c.get("nonce") ?? "", ...contexto }),
    status as any,
  );
}

/** O `RedirectResponse(..., 303)`. */
export function redirecionar(c: Ctx, destino: string, status: 301 | 302 | 303 | 307 = 303): Response {
  return c.redirect(destino, status);
}

/**
 * Devolve ao navegador a marca que a página mandou junto com o pedido de
 * download (`?baixar=<token>`), para o `csso.js` religar o botão.
 */
export function marcarDownload(c: Ctx): void {
  const token = c.req.query("baixar") ?? "";
  if (token && /^[A-Za-z0-9]{1,64}$/.test(token)) {
    setCookie(c, "csso_baixou", token, {
      maxAge: 60,
      path: "/",
      httpOnly: false,
      sameSite: "Lax",
      secure: cookieSeguro(c.req.url),
    });
  }
}

// =====================================================================
// Paginação de lista
// =====================================================================
export const POR_PAGINA = 50;

export interface Recorte<T> {
  itens: T[];
  pagina: number;
  por_pagina: number;
  total: number;
}

export function aparar(pagina: number | undefined, total: number, porPagina: number): number {
  const ultima = Math.max(1, Math.ceil(total / porPagina));
  return Math.min(Math.max(Math.trunc(pagina || 1), 1), ultima);
}

/** Recorta uma lista JÁ MATERIALIZADA (o filtro não coube na consulta). */
export function recortar<T>(itens: T[], pagina = 1, porPagina = POR_PAGINA): Recorte<T> {
  const total = itens.length;
  const p = aparar(pagina, total, porPagina);
  const inicio = (p - 1) * porPagina;
  return { itens: itens.slice(inicio, inicio + porPagina), pagina: p, por_pagina: porPagina, total };
}

/**
 * Recorta no BANCO: quem chama passa a contagem e uma função que busca a
 * página pedida com LIMIT/OFFSET.
 */
export async function recortarConsulta<T>(
  contar: () => Promise<number>,
  buscar: (limite: number, deslocamento: number) => Promise<T[]>,
  pagina = 1,
  porPagina = POR_PAGINA,
): Promise<Recorte<T>> {
  const total = await contar();
  const p = aparar(pagina, total, porPagina);
  const itens = await buscar(porPagina, (p - 1) * porPagina);
  return { itens, pagina: p, por_pagina: porPagina, total };
}

/** O `?pagina=` da URL, ou 1. */
export function numeroDaPagina(c: Ctx): number {
  const bruto = c.req.query("pagina") ?? "";
  return /^\d+$/.test(bruto) ? Number(bruto) : 1;
}

/** Monta `destino?mensagem=...` codificado, o `_aviso` das rotas Python. */
export function comMensagem(destino: string, mensagem: string, chave = "mensagem"): string {
  const sep = destino.includes("?") ? "&" : "?";
  return `${destino}${sep}${chave}=${encodeURIComponent(mensagem)}`;
}
